ccgains/historic_data.py

```diff
@@ -413,5 +413,246 @@ def prepare_request(self, dtime):
+        kline_interval = '1m'
+        interval_ms = pd.Timedelta(kline_interval).total_seconds() * 1000
+        req_params = {
+            'symbol': self.currency_pair,
+            'interval': kline_interval,
+            'limit': self.max_trades_per_query
+        }
+
+        chunk_delta = int(interval_ms * self.max_trades_per_query)
+        chunk_start = start
+        remaining_time = pd.Timedelta((end - start), 'ms')
+
+        df = pd.DataFrame()
+
+        while remaining_time > pd.Timedelta(0):
+            # Set up to get the next chunk
+            chunk_end = min(end, chunk_start + chunk_delta)
+
+            req_params.update({'startTime': chunk_start, 'endTime': chunk_end})
+            response = self._call_api(req_params)
+
+            klines = pd.DataFrame(
+                response.json(),
+                columns=[
+                    'OpenTime',
+                    'Open', 'High', 'Low', 'Close', 'Volume',
+                    'CloseTime',
+                    'QuoteAssetVolume', 'NumTrades',
+                    'TakerBuyBaseAssetVolume', 'TakerBuyQuoteAssetVolume',
+                    'Ignore'])
+            # Drop the columns we don't need
+            klines = klines.iloc[:, [4, 5, 6]]
+            # Index by interval close time and add to previous results
+            klines.CloseTime = pd.to_datetime(klines.CloseTime, unit='ms')
+            klines = klines.astype({'Close': float, 'Volume': float})
+            klines.set_index('CloseTime', inplace=True)
+            df = df.append(klines)
+
+            # Set up for next loop (if needed). If less than limit klines were
+            # returned, that means we got all we needed so OK to be negative
+            remaining_time -= self.max_trades_per_query * pd.Timedelta(kline_interval)
+            chunk_start = chunk_end
+
+            log.info('Fetched historical price data with request: %s', response.url)
+
+        # Get weighted prices, resampled with interval:
+        # (this will only return one column, the weighted prices;
+        # the total volume won't be needed anymore)
+        data = resample_weighted_average(
+            df, self.interval, 'Close', 'Volume')
+
+        # In case the data has been upsampled (some events
+        # being more separated than interval), the resulting
+        # Series will have some NaNs. Forward-fill them
+        # with the last prices before:
+        data.ffill(inplace=True)
+
+        if data.index.tzinfo is None:
+            data.index = data.index.tz_localize('UTC')
+
+        return len(df), data
+
+    def _call_api(self, params):
+        """Call Binance API with *params*, validate the response,
+        and return the results
+        """
+
+        # Be nice to the API
+        self._wait_if_needed()
+
+        # Make the API call
+        try:
+            url = self.url + self.command
+            req = requests.get(url, params=params)
+            self.last_query_time = now
```

Comment:
Shouldn't this be `pd.Timestamp.now()` instead of simply `now`?

Reply:
Whoops. Last-minute refactoring, but that's a poor excuse.

Fixed, and thanks for a good eye. Still, that was embarrassing...

Reply:
Well, no worries, those things happen. :-) We do really need more test cases and automatic testing... ;-)